### PR TITLE
"Remove Link" option in the right-click context menu for links

### PR DESCRIPTION
**Pull request by NMazzy**
## Description
#5046  Adds a "Remove Link" option in the right-click context menu for links, which when clicked would remove the URL and associated markup from the markdown text, leaving only the text.

## Changes
- Added a "menu.remove_link" field in the link right click context menu
- Created remove-markdown-link.ts file in markdown-editor/util

## Additional information

Tested on: 
- macOS Sonoma 14.1
- Ubuntu 22.04

### CODE REVIEW

**Comment by nathanlesage** [comment]
Thank you! Looks good so far -- just one comment: instead of slicing the entire node, you can also just search for the various child nodes that it should have. If I remember correctly, the link title that you want is not represented by its own node, instead you would have to check for two CodeMark nodes, and slice from the end of the first to the start of the second. That makes it more resilient to edge cases and leverages the parser more. I think I already implemented that snippet in the markdown AST parser module. I think searching for `case 'Link'` might turn it up

**Comment by MWBruce** [comment]
As suggested this has been implemented as suggested.

**Comment by nathanlesage** [comment]
Looks good. I found one issue while testing this out, though: It doesn't work with angle brackets links, i.e.:

```md
This is some text with a plain link: <https://www.google.com>
```

Right-clicking that link will give you the option to remove a link, but it won't actually remove the angled brackets as I would expect it.

**Comment by NMazzy** [comment]
> Looks good. I found one issue while testing this out, though: It doesn't work with angle brackets links, i.e.:
> 
> ```
> This is some text with a plain link: <https://www.google.com>
> ```
> 
> Right-clicking that link will give you the option to remove a link, but it won't actually remove the angled brackets as I would expect it.

I didn't even know this was a possible markdown link :)
This latest commit handles this case now.

File: source/common/modules/markdown-editor/util/remove-markdown-link.ts
Comment (nathanlesage) [inline source/common/modules/markdown-editor/util/remove-markdown-link.ts:47]: Can you be certain that the first and last characters will always be `<` or `>`, respectively? There may be other types of links that the parser understands that we don't yet account for.

File: source/common/modules/markdown-editor/util/remove-markdown-link.ts
Comment (nathanlesage) [inline source/common/modules/markdown-editor/util/remove-markdown-link.ts:49]: You mean `[]()`, not `[]<>`, right? The latter isn't even a valid link type

File: source/common/modules/markdown-editor/util/remove-markdown-link.ts
Comment (nathanlesage) [inline source/common/modules/markdown-editor/util/remove-markdown-link.ts:35]: What exactly is the purpose of this specific check …? Did you verify what the syntax tree looks like for all types of links? The passed `node` should always be some type of link, correct …?

File: source/common/modules/markdown-editor/util/remove-markdown-link.ts
Comment (nathanlesage) [inline source/common/modules/markdown-editor/util/remove-markdown-link.ts:7]: Add your name/GitHub handle/nickname here

File: source/common/modules/markdown-editor/util/remove-markdown-link.ts
Comment (nathanlesage) [inline source/common/modules/markdown-editor/util/remove-markdown-link.ts:29]: Reminds me that there's a third type: Reference style: https://spec.commonmark.org/0.31.2/#link-reference-definitions

I am not exactly sure how Codemirror handles these, but with a bunch of console.logs that should be researchable

**Review by nathanlesage — CHANGES_REQUESTED** [review]
We're getting there! Just a few more things!

File: source/common/modules/markdown-editor/util/remove-markdown-link.ts
Comment (NMazzy) [reply source/common/modules/markdown-editor/util/remove-markdown-link.ts:35]: No actually I found that there are 3 different cases of how a user can enter the right click context menu for a link:

For a [text](url) link:

1. They can right click on the [text] node which has type 'Link'
2. They can right click on the (url) node which has type 'Url'

For a <url> link:

3. They can right click on the <url> node which has type 'Url'

For cases 1. and 2. we extract the text from the 'Link' node, for case 3. we extract the text from the 'Url' node. I think the way which this was handled was a bit convoluted, I'll try simplifying this.

File: source/common/modules/markdown-editor/util/remove-markdown-link.ts
Comment (NMazzy) [reply source/common/modules/markdown-editor/util/remove-markdown-link.ts:47]: I found an existing function: shortenUrlVisually (/source/common/util/shorten-url-visually.ts) which handles what I am doing here, should I utilize this function instead?

File: source/common/modules/markdown-editor/util/remove-markdown-link.ts
Comment (NMazzy) [reply source/common/modules/markdown-editor/util/remove-markdown-link.ts:29]: I will look into this too.

Comment (NMazzy) [reply source/common/modules/markdown-editor/util/remove-markdown-link.ts]: image

At the moment the right click context menu doesn't appear when you right click on the [text][label] part of the reference style links. Is this something you want us to implement too?

Comment (MWBruce) [reply source/common/modules/markdown-editor/util/remove-markdown-link.ts]: It appears that a bug may be preventing the context menu from displaying for [Text 1][link1]. Additionally, the "remove link" option is currently visible for [link1]: https://example.com/. Nathan, do you have any suggestions on how we should handle this?

From my perspective, once the context menu bug for [Text 1][link1] is resolved, the "remove link" functionality should operate correctly and I think we should possibly raise an issue about this (which we will resolve next). However, the option to remove the link for [link1]: https://example.com/ will still be available. Could you please clarify how you would like us to proceed with this issue?

Comment (MWBruce) [reply source/common/modules/markdown-editor/util/remove-markdown-link.ts]: It appears that a bug due to the getURLForNode function call in the linkImageMenu may be preventing the context menu from displaying for [Text 1][link1] as it throws an error because of it not being able to find a URL child. Additionally, the "remove link" option is currently visible for [link1]: https://example.com/. Nathan, do you have any suggestions on how we should handle this?

From my perspective, once the context menu bug for [Text 1][link1] is resolved, the "remove link" functionality should operate correctly and I think we should possibly raise an issue about this (which we will resolve next). However, the option to remove the link for [link1]: https://example.com/ will still be available. Could you please clarify how you would like us to proceed with this issue?

File: source/common/modules/markdown-editor/util/remove-markdown-link.ts
Comment (nathanlesage) [reply source/common/modules/markdown-editor/util/remove-markdown-link.ts]: Please don't — `shortenUrlvisually` should only be used in contexts where what it spits out doesn't have to be a functioning url. As it says in the name: It's for visual purposes. Please just copy over whatever you need from that function here to prevent this from breaking.

File: source/common/modules/markdown-editor/util/remove-markdown-link.ts
Comment (nathanlesage) [reply source/common/modules/markdown-editor/util/remove-markdown-link.ts]: Ah, got it, thanks!

File: source/common/modules/markdown-editor/util/remove-markdown-link.ts
Comment (nathanlesage) [reply source/common/modules/markdown-editor/util/remove-markdown-link.ts]: I think those references … mh, it seems like we'd have to actually go searching in the syntax tree for the matching references before being able to decide whether to actually show a context menu. I would say this is out of scope for this specific issue, and should be tackled separately, especially since it needs to be fixed within a set of functionality, because iirc I never implemented reference links at any point …

Comment (MWBruce) [reply source/common/modules/markdown-editor/util/remove-markdown-link.ts]: Would you like us to report an issue for it?

**Comment by NMazzy** [comment]
Hi @nathanlesage, I believe all your comments above, except the one regarding the reference style links have been implemented.

**Comment by MWBruce** [comment]
@NMazzy I think it might be @kyaso that is reviewing PR's at the moment.

File: source/common/modules/markdown-editor/util/remove-markdown-link.ts
Comment (nathanlesage) [reply source/common/modules/markdown-editor/util/remove-markdown-link.ts]: Yes please do so!

Comment (MWBruce) [reply source/common/modules/markdown-editor/util/remove-markdown-link.ts]: Yes, done #5142

**Comment by MWBruce** [comment]
@kyaso Further to a discord conversation I had with Nathan could you please advise if there is anything further required.

**Comment by NMazzy** [comment]
> Sorry for the late reply!
> 
> I just tested your PR locally, and it seems to do its job, so everything fine code-wise from my side.
> 
> One thing we should add is a Changelog entry (see https://github.com/Zettlr/Zettlr/blob/develop/CHANGELOG.md). You can add it under the "GUI and Functionality" section.
> 
> Once this has been done, I guess, @nathanlesage, we can merge?

Hi, a changeling entry has now been added under the "GUI and Functionality" section.

**Comment by nathanlesage** [comment]
Looks good for me, too, but the linter is complaining about an extra semicolon and about another issue (see the unit test) — once that's fixed we can merge. Thanks @kyaso for the code review!

**Comment by NMazzy** [comment]
> Looks good for me, too, but the linter is complaining about an extra semicolon and about another issue (see the unit test) — once that's fixed we can merge. Thanks @kyaso for the code review!

Apologies should have checked this before hand, all should be fixed now.

**Comment by nathanlesage** [comment]
All good, no worries! I'll wait for the unit tests to finish and then merge it in! Thanks again for the work!

**Pull request merged**
